### ISSUE TITLE
Question: Canceling sagas from "upstream"

### ISSUE THREAD

**Issue by nhusher** [yelouafi/redux-saga#29]
This is almost the inverse of [this issue](https://github.com/yelouafi/redux-saga/issues/14).

In summary, I have a project where the user can generate an entity called an audience. It can take several seconds for the audience to be created through a series of asynchronous steps, only one audience can be in view at a time, and they may create several audiences in series.

More concretely: the user has a large set of controls that they can interact with to generate a new audience. Any time they manipulate the controls, a new audience is requested from the server, but the controls remain "unlocked" for the user to further refine their input and generate subsequent audiences. I need to ensure that the last audience they requested is the one being shown, but I need to be able to generate the entities in parallel. Here's what I've come up with (somewhat simplified):

```js
function* createAudienceSaga (getState) {
  let action, cancelLastAudienceBuild = () => {};

  // When a new audience create action comes along, cancel the previous one
  // by resolving the currently-active promise. We're using a promise here
  // so that `race(...)` calls can be our mechanism for determining cancelation
  /*eslint no-cond-assign:0 */
  while (action = yield take(CREATE_AUDIENCE)) {
    cancelLastAudienceBuild();

    let transaction = new Promise(r => cancelLastAudienceBuild = r);

    yield fork(buildAudience, transaction, action.payload);
  }
}

// Create a new audience
function* buildAudienceFromPlaceId(tx, params) {

  // Step 1: retrieve the place that the user wants to build the audience in
  const { place, canceled_0 } = yield race({
    place: call(api.places.get, params.placeId),
    canceled_0: tx
  });

  // If the user has fiddled with the UI, we are generating a new audience
  // in parallel and we can bail out of this one before doing any more work
  if(canceled_0) { return; }

  // Update the place information in the app state ASAP to show the user:
  yield put(showPlace(place));

  // Step 2: Create a new audience with the provided definition
  // This does not create the audience, but rather returns the audience ID,
  // which we have to request separately. The creation process is usually
  // very fast, but the step 3 can take several seconds.
  const { audienceId, canceled_1 } = yield race({
    audienceId: call(createAudienceWithDefinition, {
      definition: params.definition,
      geography: [makePlaceGeography(place)],
      product_id: params.product
    }),
    canceled_1: tx
  });

  // Again, check if a new transaction has begun before continuing:
  if(canceled_1) { return; }

  // Step 3: Retrieve the new audience data for display
  // As mentioned, this step can take several seconds for certain parameters.
  const { audience, canceled_2 } = yield race({
    audience: call(api.audience.get, audienceId),
    canceled_2: tx
  });

  // Last time, check to see if the user has changed the controls before
  // dispatching the new audience to the app state:
  if(canceled_2) { return; }

  yield put(showAudience(audience));
}
```

Yes, I know this is a pathological use case. I can't fix it, but I'm looking for a straightforward way to box this complexity up and not have to think about it.

It would be nice to have a feature on tasks that lets you cancel them from the outside. I don't know how that would work in practice without altering the API for tasks pretty significantly, or altering the internals of the saga state machine (which I haven't looked at closely and don't fully understand).

My other thought is that there's a simple and obvious way to do this that I'm not seeing.

**Comment by nhusher** [comment]
It occurs to me that I should be using the built-in exception system:

```js
function* createAudienceSaga (getState) {
  let action, cancelLastAudienceBuild = () => {};

  // When a new audience create action comes along, cancel the previous one
  // by resolving the currently-active promise. We're using a promise here
  // so that `race(...)` calls can be our mechanism for determining cancelation
  /*eslint no-cond-assign:0 */
  while (action = yield take(CREATE_AUDIENCE)) {
    cancelLastAudienceBuild();

    let transaction = new Promise(r => cancelLastAudienceBuild = r);

    yield fork(buildAudience, transaction, action.payload);
  }
}

// Create a new audience
function* buildAudienceFromPlaceId(tx, params) {
  try {
    let [place] = yield [call(api.places.get, id), tx];
    yield put(showPlace(place));
    
    const [ audienceId ] = [
      call(createAudienceWithDefinition, {
        definition: params.definition,
        geography: [makePlaceGeography(place)],
        product_id: params.product
      }), tx ];
  
    const [ audience ] = yield [call(api.audience.get, audienceId), tx ];
    
    yield put(showAudience(audience));    
  } catch (e) {
    if (e instanceof TransactionCancelledError) {
      console.info("Transaction cancelled", id, product);
      return;
    } else {
      throw e;
    }
  }
}
```

This resolves the issue to my satisfaction.

**Comment by yelouafi** [comment]
In your 2nd solution, this code

```javascript
let [place] = yield [call(api.places.get, id), tx];
```

If `call(api.places.get, id)` resolves first, it'll have to wait also for `tx` to complete (which always rejects in your code)

The task cancellation would handle this case in a more concise way, until #26 is merged, I'd also use the builtin exception system, but would opt instead for a more declarative way

```javascript
// a higher order function
function buildTx(cancelEffect) {
  // returns a reusable guarded call subroutine
  // will throw if cancelEffect completes before call(fn, ...args)
  return function* guardedCall(fn, ...args) {
    const {result} = yield race({
      result: call(fn, ...args),
      cancelEffect
    })

    if(result)
       return result
    throw new TransactionCancelledError()
  }
}
  

function* createAudienceSaga (getState) {
  let action
  const nextCreateAudience = take(CREATE_AUDIENCE)

  /*eslint no-cond-assign:0 */
  while (action = yield nextCreateAudience ) {
    yield fork(
        buildAudience, 
        nextCreateAudience,
        action.payload);
  }
}

// Create a new audience
function* buildAudienceFromPlaceId(cancelEffect, params) {
  const tx = buildTx(cancelEffect)
  try {
    let place = yield call(tx, api.places.get, id);
    yield put(showPlace(place));

    const audienceId = yield
      call(tx, createAudienceWithDefinition, {
        definition: params.definition,
        geography: [makePlaceGeography(place)],
        product_id: params.product
      });

    const audience = yield call(tx, api.audience.get, audienceId);

    yield put(showAudience(audience));    
  } catch (e) {
    if (e instanceof TransactionCancelledError) {
      console.info("Transaction cancelled", id, product);
      return;
    } else {
      throw e;
    }
  }
}
```

**Comment by nhusher** [comment]
Thanks! I'll give this a look over and attempt to grok it.

**Comment by nhusher** [comment]
I ended up taking your higher-order function and making it lower-order. This makes it easier to test:

```js
export function* guard(tx, ...args) {
  const { result } = yield race({
    result: call(...args),
    tx
  });

  if(result) {
    return result;
  } else {
    throw new TransactionCancelledError();
  }
}
```

```
// Create a new audience from a place and product id
export function* buildAudienceFromPlaceId(tx, id, product) {
  try {
    // Step 1: retrieve the place that the user wants to build the audience in
    let place = yield call(guard, tx, api.places.get, id);
    // ...
```

It obscures intent somewhat because the actual function being called is deep in the arguments list, but it makes the test very straightforward:

```
import { guard } from 'wherever';

let tx = new Promise(() => {}),
    iterator = buildAudienceFromPlaceId(tx);

assert.deepEqual(
    iterator.next().value,
    call(guard, tx, api.places.get, placeId));
```

**Comment by yelouafi** [comment]
I see, you can also define a higher order effect creator

```javascript
export function* guard(tx, ...args) {...}

const guardedCall = (guard, tx) => call(guard, tx, api.places.get, id);

// Create a new audience from a place and product id
export function* buildAudienceFromPlaceId(tx, id, product) {
  const callWithTx= guardedCall(guard, tx)
  try {
    // Step 1: retrieve the place that the user wants to build the audience in
    let place = yield callWithTx(api.places.get, id);
    // ...
```

**Comment by nhusher** [comment]
I had also considered (but have not tested) the following because it uses private values that might break:

```js
// A saga that waits for CREATE_AUDIENCE actions and uses the data to construct and load a new audience:
export default function* (getState) {
  let action, lastAction;

  while (action = yield take(CREATE_AUDIENCE)) {
    if(lastAction && lastAction.isRunning()) {
      lastAction._generator.throw(new TransactionCancelledError());
    }
    lastAction = yield fork(buildAudience, action.payload);
  }
}
```

Not sure what the implications of that are.

**Comment by nhusher** [comment]
> I see, you can also define a higher order effect creator

Yeah. Again, it means peeling the function instance out of the `iterator.next().value`, which I'd rather avoid if possible. If I use a lower-order function, I can share the function more overtly between the application code and tests, at the cost of somewhat obscured intent. Either way has tradeoffs, and I might change my mind later.

**Comment by yelouafi** [comment]
>I had also considered (but have not tested) the following because it uses private values that might break

I wouldn't recommend this. As internal representation may change (actually already changed in the current master branch).

**Comment by yelouafi** [comment]
>Yeah. Again, it means peeling the function instance out of the iterator.next().value

Maybe I missed something, but is there any issue with the following

code
```javascript
export function* guard(tx, ...args) {...}

export const guardedCall = (guard, tx) => call(guard, tx, api.places.get, id);

// Create a new audience from a place and product id
export function* buildAudienceFromPlaceId(tx, id, product) {
  const callWithTx= guardedCall(guard, tx)
  try {
    // Step 1: retrieve the place that the user wants to build the audience in
    let place = yield callWithTx(api.places.get, id);
    // ...
```

test
```javascript
import { guard, guardedCall } from 'wherever';

let tx = new Promise(() => {}),
     callWithTx = guardedCall(guard, tx),
     iterator = buildAudienceFromPlaceId(tx);

assert.deepEqual(
    iterator.next().value,
    callWithTx(api.places.get, placeId));
```

**Comment by nhusher** [comment]
The result of a higher-order function is always unique:

```
function add(a) { return function(b) { return a + b; } }
assert(add(5) === add(5)); // untrue, fails
```

This means that the call to `guardedCall` in the generator won't equal the equivalent call in the test.

**Comment by yelouafi** [comment]
>The result of a higher-order function is always unique:

yes, but in our current case we're not comparing the resulting functions but the final value

```javascript
function add(a) { return function(b) { return a + b; } }
assert(add(5)(3) === add(5)(3)); 

assert.deepEqual(
  guardedCall(guard, tx)((api.places.get, placeId),
  guardedCall(guard, tx)((api.places.get, placeId)
)
```

**Comment by yelouafi** [comment]
It's not explicit in the docs, but all effect creators like `call` are pure functions. for example a call returns an object {fn, args}, so as long you're providing the same inputs, which is the case above for all the 4 arguments, you can expect the same output (in the deepEqual sens)

**Comment by nhusher** [comment]
Oh yeah. I knew that but wasn't really understanding what was going on. Concern retracted.

Thanks for all your help, by the way!
